src/acl.c

```diff
@@ -135,6 +135,7 @@ typedef struct {
                         field is NULL the user cannot mention any channel in a
                         `PUBLISH` or [P][UNSUBSCRIBE] command, unless the flag
                         ALLCHANNELS is set in the user. */
+    sds categories; /* A string representation of the ordered categories */
```

Comment:
> The whole point of storing the raw ACL is that we can re-apply is in case a module is loaded at runtime (which is something we do support can can't put aside IMO), so if we do that we also want to let people enable a command that doesn't (yet) exist.

I think we only really care about categories in this case, specifically the case where module commands register themselves to existing categories. If a module creates a new command or category, I'm okay with asking them to then update the ACLs on their users to include the new commands/categories. I think there is a risk someone will provide an incorrect ACL but think it applied cleanly (the error case you mentioned). I think the validation here is important, as you mentioned. Missing a `+` means that the user will not have the permissions the admin expected, and missing a `-` means that there are permissions that should be excluded but aren't. I think coming up with a non-fatal error is risky, as it might be skipped or ignored by existing systems. 

> i'd like it to return to the same state if the module is later re-loaded.

I think the real missing link here is that we don't natively support "reloading" or "updating" a module. There are so many issues with unloading where we can't easily cleanup the state.

Reply:
@yossigo Would appreciate your thoughts about if we need to retain the entire ACL string or not.

Reply:
@madolson, I agree that unloading or reloading a module is not a mainstream flow, and we'd be better off breaking it rather than be too soft on error-checking. But what are the arguments against storing the entire ACL string and avoiding the exceptions you listed above? It seems more trivial to me.

Reply:
@yossigo The simplest option is to just store the entire string verbatim, and concatenate new fields on to it. We could do that. I think it is useful to compact the string where applicable though, at that point I don't know the simplicity really matters and the question of how much compaction we do becomes relevant.

Reply:
The only example I can think of is after an update it can be a lot more readable. Something like "+@all reset -@all" is just ugly to retain, since the ending "-@all" is clearest. 

Also worth calling out, we have to at least process it, since we need to strip out passwords and convert to hashes and what not. We need not keep "everything".

Reply:
Circling back to this, I think that last comment makes sense. So, we store verbatim strings but apply some exceptions:
* Redaction of passwords, etc.
* Compaction that's obvious and has no potential future loss of information (e.g., safe to truncate anything before a reset, assuming we won't change reset behavior).

Reply:
Thanks for circling back. Let me propose the following compaction rules then, which are aligned with the tenets:
* Redact passwords.
* -@all and +@all will truncate everything before.
* duplicate and/or inverse rules (+write and -write) will compact into their latest values. This should be "lossless". 

I think this covers all of the use cases we identified.